File: scripts/seed-supabase.ts
/**
 * Seed script for Supabase PostgreSQL
 * Usage: DATABASE_URL="postgresql://..." npx tsx scripts/seed-supabase.ts
 */
import * as XLSX from 'xlsx';
import { PrismaClient } from '@prisma/client';
import * as crypto from 'crypto';
import * as path from 'path';

const prisma = new PrismaClient();

function hashPassword(password: string): string {
  return crypto.createHash('sha256').update(password).digest('hex');
}

async function seed() {
  console.log('Connecting to database...');
  await prisma.$connect();
  console.log('Connected!');

  console.log('Seeding divisions...');
  const divisions = await Promise.all([
    prisma.division.upsert({ where: { name: 'TSA' }, update: {}, create: { name: 'TSA', color: '#3b82f6' } }),
    prisma.division.upsert({ where: { name: 'JLM' }, update: {}, create: { name: 'JLM', color: '#10b981' } }),
    prisma.division.upsert({ where: { name: 'AMT' }, update: {}, create: { name: 'AMT', color: '#f59e0b' } }),
    prisma.division.upsert({ where: { name: 'NOSS' }, update: {}, create: { name: 'NOSS', color: '#ef4444' } }),
    prisma.division.upsert({ where: { name: 'Konstruksi' }, update: {}, create: { name: 'Konstruksi', color: '#8b5cf6' } }),
  ]);

  console.log('Seeding users...');
  await prisma.user.upsert({
    where: { username: 'admin' },
    update: {},
    create: { username: 'admin', name: 'Administrator', password: hashPassword('asrama33'), role: 'ADMIN' }
  });

  const editorUsers = [
    { username: 'tsa_editor', name: 'Editor TSA', divisionName: 'TSA' },
    { username: 'jlm_editor', name: 'Editor JLM', divisionName: 'JLM' },
    { username: 'amt_editor', name: 'Editor AMT', divisionName: 'AMT' },
    { username: 'noss_editor', name: 'Editor NOSS', divisionName: 'NOSS' },
    { username: 'konstruksi_editor', name: 'Editor Konstruksi', divisionName: 'Konstruksi' },
  ];

  for (const eu of editorUsers) {
    const div = divisions.find(d => d.name === eu.divisionName)!;
    await prisma.user.upsert({
      where: { username: eu.username },
      update: {},
      create: { username: eu.username, name: eu.name, password: hashPassword('password123'), role: 'EDITOR', divisionId: div.id }
    });
  }

  await prisma.user.upsert({
    where: { username: 'viewer' },
    update: {},
    create: { username: 'viewer', name: 'Viewer', password: hashPassword('viewer123'), role: 'VIEWER' }
  });

  // Import Excel data
  const excelPaths = [
    path.join(process.cwd(), 'upload/138 LIST OUTLIERS TSA 11 AGUSTUS 2026 (1).xlsx'),
    path.join(process.cwd(), '../upload/138 LIST OUTLIERS TSA 11 AGUSTUS 2026 (1).xlsx'),
  ];

  let excelPath = excelPaths.find(p => {
    try { require('fs').accessSync(p); return true; } catch { return false; }
  });

  if (excelPath) {
    console.log(`Importing Excel data from ${excelPath}...`);
    const workbook = XLSX.readFile(excelPath);
    const ws = workbook.Sheets['OUTLIERS TSA'];
    const rows = XLSX.utils.sheet_to_json(ws, { header: 1 }) as any[][];

    let count = 0;
    for (let i = 2; i < rows.length; i++) {
      const r = rows[i];
      if (!r[0] || r[0] === null) continue;
      const orderNum = typeof r[0] === 'number' ? r[0] : parseInt(String(r[0])) || (i - 1);
      const homepass = typeof r[9] === 'number' ? r[9] : parseInt(String(r[9])) || 0;
      const odp = typeof r[10] === 'number' ? r[10] : parseInt(String(r[10])) || 0;
      const indexNum = typeof r[8] === 'number' ? r[8] : parseInt(String(r[8])) || 0;

      await prisma.monitoringRow.create({
        data: {
          orderNum, categoryBak: String(r[1] || ''), provinsi: String(r[2] || ''),
          kabupaten: String(r[3] || ''), kecamatan: String(r[4] || ''),
          kelurahan: String(r[5] || ''), kelRwSiteName: String(r[6] || ''),
          desaPerum: String(r[7] || ''), indexNum, homepass, odp,
          remarksTsa: String(r[11] || ''), klasifikasiTsa: String(r[12] || ''),
          picTsa: String(r[13] || ''), remarksJlm: String(r[14] || ''),
          customData: '{}',
        }
      });
      count++;
    }
    console.log(`Imported ${count} rows from Excel`);
  } else {
    console.log('Excel file not found. Skipping data import.');
    console.log('You can import data manually via the app admin panel.');
  }

  console.log('Creating chart configs...');
  const chartTypes = ['bar', 'line', 'pie', 'area', 'bar', 'line', 'pie', 'bar'];
  const existingCharts = await prisma.chartConfig.count();
  if (existingCharts === 0) {
    for (let i = 0; i < 8; i++) {
      await prisma.chartConfig.create({
        data: { title: `Chart ${i + 1}`, chartType: chartTypes[i], config: JSON.stringify({}), order: i }
      });
    }
    console.log('Created 8 chart configs');
  }

  console.log('\nSeeding complete!');
  await prisma.$disconnect();
}

seed().catch(e => { console.error(e); process.exit(1); });

File: scripts/deploy/seed-supabase.py
import subprocess, json, sys, os

token = os.environ.get('SUPABASE_ACCESS_TOKEN', '')
if not token:
    print('Error: Set SUPABASE_ACCESS_TOKEN environment variable')
    sys.exit(1)
url = 'https://api.supabase.com/v1/projects/bjjhswrpsmvljqsbbycj/database/query'

def run_sql(sql, desc='SQL'):
    payload = json.dumps({'query': sql})
    result = subprocess.run(
        ['curl', '-s', '-X', 'POST',
         '-H', f'Authorization: Bearer {token}',
         '-H', 'Content-Type: application/json',
         '-d', payload, url],
        capture_output=True, text=True, timeout=120
    )
    if result.returncode != 0:
        print(f'  {desc}: CURL ERROR - {result.stderr[:100]}')
        return None
    try:
        data = json.loads(result.stdout)
        if isinstance(data, list):
            print(f'  {desc}: OK ({len(data)} rows)')
        else:
            print(f'  {desc}: {data}')
        return data
    except:
        print(f'  {desc}: RAW - {result.stdout[:100]}')
        return None

# Step 1: Disable RLS
print('=== Disabling RLS ===')
for tbl in ['Division', 'User', 'MonitoringRow', 'CustomColumn', 'AuditLog', 'ChartConfig']:
    run_sql(f'ALTER TABLE "{tbl}" DISABLE ROW LEVEL SECURITY;', f'Disable RLS {tbl}')

# Step 2: Clean test data
print('\n=== Cleaning test data ===')
run_sql("DELETE FROM \"Division\" WHERE \"id\" = 'test-id';", 'Clean test')

# Step 3: Seed divisions
print('\n=== Seeding Divisions ===')
divisions_sql = '''
INSERT INTO "Division" ("id", "name", "color") VALUES (gen_random_uuid()::text, 'TSA', '#3b82f6') ON CONFLICT ("name") DO NOTHING;
INSERT INTO "Division" ("id", "name", "color") VALUES (gen_random_uuid()::text, 'JLM', '#10b981') ON CONFLICT ("name") DO NOTHING;
INSERT INTO "Division" ("id", "name", "color") VALUES (gen_random_uuid()::text, 'AMT', '#f59e0b') ON CONFLICT ("name") DO NOTHING;
INSERT INTO "Division" ("id", "name", "color") VALUES (gen_random_uuid()::text, 'NOSS', '#ef4444') ON CONFLICT ("name") DO NOTHING;
INSERT INTO "Division" ("id", "name", "color") VALUES (gen_random_uuid()::text, 'Konstruksi', '#8b5cf6') ON CONFLICT ("name") DO NOTHING;
SELECT count(*) as divisions FROM "Division";
'''
run_sql(divisions_sql, 'Divisions')

# Step 4: Seed users
print('\n=== Seeding Users ===')
import hashlib
def h(pw): return hashlib.sha256(pw.encode()).hexdigest()

users_sql = f'''
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'admin', 'Administrator', '{h("asrama33")}', 'ADMIN', NULL, NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'tsa_editor', 'Editor TSA', '{h("password123")}', 'EDITOR', (SELECT "id" FROM "Division" WHERE "name" = 'TSA'), NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'jlm_editor', 'Editor JLM', '{h("password123")}', 'EDITOR', (SELECT "id" FROM "Division" WHERE "name" = 'JLM'), NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'amt_editor', 'Editor AMT', '{h("password123")}', 'EDITOR', (SELECT "id" FROM "Division" WHERE "name" = 'AMT'), NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'noss_editor', 'Editor NOSS', '{h("password123")}', 'EDITOR', (SELECT "id" FROM "Division" WHERE "name" = 'NOSS'), NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'konstruksi_editor', 'Editor Konstruksi', '{h("password123")}', 'EDITOR', (SELECT "id" FROM "Division" WHERE "name" = 'Konstruksi'), NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
INSERT INTO "User" ("id", "username", "name", "password", "role", "divisionId", "createdAt", "updatedAt") VALUES (gen_random_uuid()::text, 'viewer', 'Viewer', '{h("viewer123")}', 'VIEWER', NULL, NOW(), NOW()) ON CONFLICT ("username") DO NOTHING;
SELECT count(*) as users FROM "User";
'''
run_sql(users_sql, 'Users')

# Step 5: Seed monitoring rows
print('\n=== Seeding Monitoring Rows ===')
import openpyxl
wb = openpyxl.load_workbook('/home/z/my-project/upload/138 LIST OUTLIERS TSA 11 AGUSTUS 2026 (1).xlsx', read_only=True)
ws = wb['OUTLIERS TSA']
rows = []
for i, row in enumerate(ws.iter_rows(values_only=True)):
    if i < 2: continue
    if row[0] is None: continue
    rows.append(list(row))
wb.close()

print(f'  Read {len(rows)} rows from Excel')

# Build and send in batches of 20
for batch_start in range(0, len(rows), 20):
    batch = rows[batch_start:batch_start+20]
    stmts = []
    for idx, r in enumerate(batch):
        i = batch_start + idx
        orderNum = r[0] if isinstance(r[0], (int,float)) else (int(float(str(r[0]))) if str(r[0]).strip().replace('.','').replace('-','').isdigit() else i+1)
        vals = [f'gen_random_uuid()::text', str(int(orderNum))]
        for j in range(1, 15):
            v = str(r[j]) if j < len(r) and r[j] is not None else ''
            v = v.replace("'", "''")
            if j in [8, 9, 10]:
                try: vals.append(str(int(float(v))))
                except: vals.append('0')
            else:
                vals.append(f"'{v}'")
        vals.append("'{}'")
        vals.append('NOW()')
        vals.append('NOW()')
        cols = '"id","orderNum","categoryBak","provinsi","kabupaten","kecamatan","kelurahan","kelRwSiteName","desaPerum","indexNum","homepass","odp","remarksTsa","klasifikasiTsa","picTsa","remarksJlm","customData","createdAt","updatedAt"'
        stmts.append(f'INSERT INTO "MonitoringRow" ({cols}) VALUES ({", ".join(vals)});')
    sql = '\n'.join(stmts) + f'\nSELECT {len(batch)} as batch_{batch_start//20+1};'
    run_sql(sql, f'Rows {batch_start+1}-{batch_start+len(batch)}')

# Step 6: Seed chart configs
print('\n=== Seeding Chart Configs ===')
charts_sql = '''
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 1', 'bar', '{}', 0);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 2', 'line', '{}', 1);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 3', 'pie', '{}', 2);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 4', 'area', '{}', 3);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 5', 'bar', '{}', 4);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 6', 'line', '{}', 5);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 7', 'pie', '{}', 6);
INSERT INTO "ChartConfig" ("id","title","chartType","config","order") VALUES (gen_random_uuid()::text, 'Chart 8', 'bar', '{}', 7);
SELECT count(*) as charts FROM "ChartConfig";
'''
run_sql(charts_sql, 'Charts')

# Step 7: Verify all data
print('\n=== Verification ===')
run_sql('SELECT (SELECT count(*) FROM "Division") as divisions, (SELECT count(*) FROM "User") as users, (SELECT count(*) FROM "MonitoringRow") as monitoring_rows, (SELECT count(*) FROM "ChartConfig") as charts', 'Final Count')
print('\nDone!')
